### ISSUE TITLE
epic(attention): witnessed attention-mass as the ground-truth context hit — the granularity ladder + rolling accumulator + closed-loop eviction

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#851]
## The thesis

When you only *consume* a model API, "what context was a hit?" is a guess — you infer it post-hoc from lexical overlap (which is exactly what `ctxplan.Outcome.Hits` does today, and only in the bench, never in production: `cmd/ctxplanbench/main.go:619`). When you **control attention itself** — the forward pass, the softmax, the KV cache — "hit" stops being inferred and becomes **witnessed**: the normalized attention weight that actually landed on a span's tokens this turn. fak owns that surface (`internal/model/forward.go:258`, `internal/kvmmu/kvmmu.go`), so it can measure the hit at the granularity attention itself operates, and — because the same kernel that measures also *evicts* — act on it in the same seam (closed-loop, the edge over an open-loop cache).

This epic turns the context signal-to-noise metric (shipped: `internal/ctxplan/signalnoise.go`, `16b8261`) from a coarse inferred boolean into a witnessed, attention-weighted, time-integrated signal — and then uses it to drive eviction.

## The formula, generalized

The shipped S/N is one rung of a ladder. The general form is fixed; only the attribution term `a` changes:

```
                 Σ_s  a_s · cost(s)          ( s ranges over resident spans )
   S/N  =  ───────────────────────────       ( cost(s) = the span's token cost )
                 Σ_s  cost(s)                 ( a_s ∈ [0,1] = attribution weight )
```

**The granularity ladder of `a_s` — inferred → witnessed → continuous → sub-span:**

| Rung | `a_s` is… | Source | Granularity | Status |
|---|---|---|---|---|
| 0 | inferred boolean `hit_s ∈ {0,1}` | lexical overlap, post-hoc | whole span | **shipped** (`Outcome.Hits`) |
| 1 | witnessed boolean `read_s ∈ {0,1}` | did the forward pass touch span `s`'s KV at all | whole span | this epic |
| 2 | **attention mass** `Σ_t w(t→s) / Σ_t w(t→·)` ∈ [0,1] | post-softmax weights, summed over generated tokens | whole span | this epic |
| 3 | per-**token** attention mass | weight per resident token, not just per span | sub-span | this epic |
| 4 | per-**head/layer** decomposition | which heads attend where | sub-span | research |

The shipped metric is rung 0: the binary special case of the witnessed one. **The formula does not change as we climb; only `a_s` gets more truthful.** That is the load-bearing design property — it means every rung is a drop-in refinement of a number that already has a consumer.

## The temporal dimension — hit is a rolling accumulation, not a per-turn event

A span's value is not decided by one turn. A tool result can idle for 10 turns then become load-bearing; or run hot early and die. So the witnessed quantity is a **per-span attention accumulator over the session**, and the reduction over time is chosen by the consumer:

```
   real-time controller:   A_s(t) = λ · A_s(t−1) + a_s(t)     ( EMA, recency-decayed )
   post-hoc analyst:       A_s     = Σ_t a_s(t)               ( undecided cumulative sum )
                           + trajectory: the {a_s(t)} series  ( when was it hot? )
```

This is the **same accumulator with two reductions**, and the decay constant λ is the *only* knob separating them:

- **Real-time (λ < 1)** — the eviction/pin controller wants "what is hot *now*". A recency-decayed rolling sum of attention mass is, structurally, the **H2O / heavy-hitter** signal — but as a *witnessed kernel quantity* the kernel can act on in the same seam, not an approximation bolted onto a rented cache. Pin high-`A_s`, evict low-`A_s`.
- **Post-hoc (λ = 1)** — the analyst wants "where did attention go across the whole run". The cumulative sum + the per-turn trajectory answers "was this 9000-token blob *ever* worth its residency, and for how many turns was it dead weight?" — the audit S/N integrated over the session.

So the metric is parameterized by `(rung, λ)`: `(0, ·)` is shipped; `(2, λ<1)` is the live attention-informed evictor; `(2, λ=1)` is the post-hoc session attention report.

## The honest novelty boundary

fak today proves **exact** eviction (`max|Δ|=0` span removal, witnessed against a HuggingFace oracle — `docs/explainers/addressable-kv-cache.md:105`) and **policy-driven** eviction (a verdict evicts, not cache pressure). It does **not** witness attention; H2O / SnapKV / Quest / PyramidKV are cited as the *unbuilt* attention-mass axis (`docs/industry-scorecard/memory.md:93`).

What this epic claims — and what it must NOT overclaim:

- **CLAIM:** fak controls the KV, so it can witness attention mass as a first-class kernel signal, attribute it to spans (`kvmmu.Segment{ID,From,Len}` already maps span→KV positions), accumulate it over a session, and make eviction **attention-informed AND exact** — the combination the lossy-attention literature (approximate) and fak-today (exact but attention-blind) each lack one half of.
- **DO NOT CLAIM:** that fak invented heavy-hitters (H2O did), or that attention-mass keying is novel. The novelty is the *witness* (a measured, replayable kernel signal, not a heuristic) and the *fusion with exact eviction*.
- **DO NOT overclaim the "kernel controls attention" metaphor** — the positioning brief (`docs/launch/positioning-brief.md:64`) warns the gate runs in-process, not behind a privilege ring. The claim is observability + closed-loop action, not a hardware boundary.

## Children (the rungs, smallest-first)

1. **Attention-mass witness hook** — emit post-softmax weights at the `softmaxInPlace` seam (the "missing 50 bytes"), behind a default-off observer, math byte-unchanged. The foundation everything else reads.
2. **Span attention attribution** — accumulate emitted weights onto `kvmmu.Segment` via the existing `{From,Len}`→`ID` map; add the `Attended` accumulator + reset-on-evict.
3. **Witnessed S/N (rung 1→2)** — feed the attribution into `ComputeSignalNoise` as the `a_s` term, replacing the inferred boolean; prove the witnessed ratio against the inferred one on a real trace.
4. **Rolling session accumulator (the temporal dual)** — the per-span EMA (real-time) + cumulative (post-hoc) reductions; one accumulator, two λ.
5. **Attention-informed exact eviction** — close the loop: the controller evicts lowest-`A_s` spans via the *existing* exact `KVCache.Evict`, so eviction is attention-informed AND `max|Δ|=0`. The H2O-meets-exactness rung.
6. **Post-hoc session attention report** — the analyst surface: per-span cumulative mass + trajectory + integrated S/N over the run, in the session audit.
7. **Production Outcome producer** — close the rung-0 loop too (today `Outcome` has no production producer); a witnessed Outcome from the attribution, so the planner learns from real hits, not lexical guesses.
8. **Determinism + replay witness** — the attention witness must be replayable (same trace → same accumulator) and must NOT perturb the forward pass; a paired honesty test proving byte-identical logits with the observer on vs off.

## Foundations this builds on (do not re-derive)

- Shipped S/N metric: `internal/ctxplan/signalnoise.go`, explainer `docs/explainers/context-signal-to-noise.md`.
- Span→KV map: `kvmmu.Segment{ID,From,Len}` (`internal/kvmmu/kvmmu.go:48`); exact eviction `KVCache.Evict(from,n)` (`internal/model/kvcache.go:94`).
- Attention compute seam: `internal/model/forward.go:258`, `batch_attn.go:64-83`, `prefill_attn.go:57`.
- The inference-front-end lens (one decision, two depths: wire-byte + KV-tensor).

## Children (live)

- [ ] #852 — rung 1: attention-mass witness hook (the "missing 50 bytes")
- [ ] #853 — rung 2: span attention attribution (`kvmmu.Segment.Attended`)
- [ ] #854 — rung 1→2: witnessed S/N (attention mass as `a_s`)
- [ ] #855 — rung 4: rolling per-span accumulator (EMA + cumulative)
- [ ] #856 — rung 5: attention-informed exact eviction (H2O ∩ exactness)
- [ ] #857 — rung 6: post-hoc session attention report
- [ ] #858 — rung 7: production Outcome producer (close the planner loop)
- [ ] #859 — rung 8: attention-witness honesty (observer-on==off, replay, conservation)

Suggested order: 852 → 853 → {854, 855} → {856, 858} → 856-dependent 857 → 856 eviction; 859 gates all "witnessed" claims and should land alongside 852.

**Comment by anthony-chaudhary** [comment]
Rung 0 (the inferred-boolean special case of this formula) is already SHIPPED: `internal/ctxplan/signalnoise.go` (origin/main `16b8261`) + explainer `docs/explainers/context-signal-to-noise.md` (`fc92dee`). This epic replaces the inferred `a_s` with the witnessed one. See the measured cache-hit-trap data (0.88→0.99 by length on 247 sessions) in that explainer.

**Comment by anthony-chaudhary** [comment]
Resolved by `c5916d2` (resolving commit). Closed by the DOS dispatch loop's close-resolved arm, witnessed via `dos commit-audit` (verdict OK / diff-witnessed). Reopen if this does not fully resolve it.

**Comment by anthony-chaudhary** [comment]
Research sibling filed: **#860 — witnessed attention attribution as a REWARD over context spans** (credit assignment, not just eviction — the use H2O left on the table). #851 builds the witness substrate (rung 1-2 already partly in-tree: `internal/model/attn_observer.go`, `signalnoise.go:SignalNoiseFromAttention`); #860 asks what that signal is WORTH as a reward, prior-art-survey-first (#861-865) before any experiment (#866-867).
